File: be/src/main/java/com/bemo/hr/helpdesk/application/HelpdeskService.java
package com.bemo.hr.helpdesk.application;

import com.bemo.hr.helpdesk.domain.*;
import com.bemo.hr.shared.domain.BusinessRuleException;
import com.bemo.hr.shared.security.TenantContext;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;

@Service
@RequiredArgsConstructor
public class HelpdeskService {

    private final HelpdeskCategoryRepository categoryRepo;
    private final TicketRepository ticketRepo;
    private final TicketMessageRepository messageRepo;

    @Transactional
    public HelpdeskCategory createCategory(String appId, String nameAr, String nameEn,
                                           int slaFirstResp, int slaResolution) {
        var cat = new HelpdeskCategory(appId, nameAr, nameEn, slaFirstResp, slaResolution);
        return categoryRepo.save(cat);
    }

    @Transactional(readOnly = true)
    public List<HelpdeskCategory> listCategories(String appId) {
        return categoryRepo.findByAppIdOrderByCreatedAtDesc(appId);
    }

    @Transactional
    public Ticket createTicket(String appId, String requesterUserId, String categoryId,
                               String title, String description, Ticket.Priority priority) {
        HelpdeskCategory cat = categoryRepo.findById(categoryId)
                .orElseThrow(() -> new BusinessRuleException("Category not found.",
                        "TICKET_CATEGORY_NOT_FOUND", HttpStatus.NOT_FOUND));
        long nextNo = ticketRepo.maxTicketNo(appId) + 1;
        Ticket ticket = new Ticket(appId, nextNo, requesterUserId, categoryId, title, description, priority);
        ticket.stampSla(Instant.now(), cat.getSlaFirstResponseHours(), cat.getSlaResolutionHours(), priority);
        return ticketRepo.save(ticket);
    }

    @Transactional(readOnly = true)
    public List<Ticket> listTickets(String appId, String status, String assigneeUserId) {
        if (status != null) return ticketRepo.findByAppIdAndStatusOrderByTicketNoDesc(appId, status);
        if (assigneeUserId != null) return ticketRepo.findByAppIdAndAssigneeUserIdOrderByTicketNoDesc(appId, assigneeUserId);
        return ticketRepo.findByAppIdOrderByTicketNoDesc(appId);
    }

    @Transactional(readOnly = true)
    public Ticket getTicket(String appId, String ticketId) {
        return ticketRepo.findById(ticketId)
                .filter(t -> t.getAppId().equals(appId))
                .orElseThrow(() -> new BusinessRuleException("Ticket not found.",
                        "TICKET_NOT_FOUND", HttpStatus.NOT_FOUND));
    }

    @Transactional
    public void assignTicket(String appId, String ticketId, String assigneeUserId) {
        Ticket t = getTicket(appId, ticketId);
        t.assign(assigneeUserId);
        ticketRepo.save(t);
    }

    @Transactional
    public void transitionTicket(String appId, String ticketId, Ticket.Status newStatus) {
        Ticket t = getTicket(appId, ticketId);
        t.transitionTo(newStatus);
        if (newStatus == Ticket.Status.RESOLVED) t.resolve(Instant.now());
        ticketRepo.save(t);
    }

    @Transactional
    public TicketMessage addMessage(String appId, String ticketId, String authorUserId,
                                    String body, boolean internalNote) {
        Ticket t = getTicket(appId, ticketId);
        TicketMessage msg = new TicketMessage(appId, ticketId, authorUserId, body, internalNote);
        messageRepo.save(msg);
        if (!internalNote && t.getFirstResponseAt() == null) {
            t.markFirstResponse(Instant.now());
            ticketRepo.save(t);
        }
        return msg;
    }

    @Transactional(readOnly = true)
    public List<TicketMessage> listMessages(String appId, String ticketId, boolean includeInternal) {
        getTicket(appId, ticketId);
        if (includeInternal) return messageRepo.findByTicketIdOrderByCreatedAtAsc(ticketId);
        return messageRepo.findByTicketIdAndInternalNoteFalseOrderByCreatedAtAsc(ticketId);
    }

    public long countOpenTickets(String appId) {
        return ticketRepo.countByAppIdAndStatus(appId, Ticket.Status.NEW.name())
             + ticketRepo.countByAppIdAndStatus(appId, Ticket.Status.OPEN.name());
    }

    public long countMyOpenTickets(String appId, String userId) {
        return ticketRepo.countByAppIdAndAssigneeUserIdAndStatusIn(appId, userId,
                List.of(Ticket.Status.NEW.name(), Ticket.Status.OPEN.name()));
    }

    @Transactional
    public void checkSlaBreaches(String appId) {
        Instant now = Instant.now();
        ticketRepo.findByAppIdOrderByTicketNoDesc(appId).stream()
            .filter(t -> t.getStatus() != Ticket.Status.CLOSED && t.getStatus() != Ticket.Status.RESOLVED)
            .forEach(t -> {
                boolean changed = false;
                if (!t.isSlaBreachFirstResponse() && t.getFirstResponseAt() == null
                        && t.getDueFirstResponse() != null && now.isAfter(t.getDueFirstResponse())) {
                    t.setSlaBreachFirstResponse(true);
                    changed = true;
                }
                if (!t.isSlaBreachResolution() && t.getResolvedAt() == null
                        && t.getDueResolution() != null && now.isAfter(t.getDueResolution())) {
                    t.setSlaBreachResolution(true);
                    changed = true;
                }
                if (changed) ticketRepo.save(t);
            });
    }
}

File: fe/src/app/features/helpdesk/helpdesk.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { firstValueFrom } from 'rxjs';
import { HelpdeskCategory, Ticket, TicketMessage, TicketListResponse, KbArticle } from './helpdesk.models';

@Injectable({ providedIn: 'root' })
export class HelpdeskService {
  private readonly http = inject(HttpClient);

  async listCategories(): Promise<HelpdeskCategory[]> {
    return firstValueFrom(this.http.get<HelpdeskCategory[]>('/api/v1/helpdesk/categories'));
  }

  async createCategory(payload: { nameAr: string; nameEn: string; slaFirstResponseHours: number; slaResolutionHours: number }): Promise<HelpdeskCategory> {
    return firstValueFrom(this.http.post<HelpdeskCategory>('/api/v1/helpdesk/categories', payload));
  }

  async listTickets(status?: string, assigneeUserId?: string): Promise<TicketListResponse> {
    const params: Record<string, string> = {};
    if (status) params['status'] = status;
    if (assigneeUserId) params['assigneeUserId'] = assigneeUserId;
    return firstValueFrom(this.http.get<TicketListResponse>('/api/v1/helpdesk/tickets', { params }));
  }

  async createTicket(payload: { categoryId: string; title: string; description: string; priority?: string }): Promise<Ticket> {
    return firstValueFrom(this.http.post<Ticket>('/api/v1/helpdesk/tickets', payload));
  }

  async getTicket(id: string): Promise<Ticket> {
    return firstValueFrom(this.http.get<Ticket>(`/api/v1/helpdesk/tickets/${id}`));
  }

  async assignTicket(id: string, assigneeUserId: string): Promise<void> {
    await firstValueFrom(this.http.post(`/api/v1/helpdesk/tickets/${id}/assign`, { assigneeUserId }));
  }

  async transitionTicket(id: string, status: string): Promise<void> {
    await firstValueFrom(this.http.post(`/api/v1/helpdesk/tickets/${id}/transition`, { status }));
  }

  async addMessage(ticketId: string, body: string, internalNote: boolean): Promise<TicketMessage> {
    return firstValueFrom(this.http.post<TicketMessage>(`/api/v1/helpdesk/tickets/${ticketId}/messages`, { body, internalNote }));
  }

  async listMessages(ticketId: string, includeInternal: boolean): Promise<TicketMessage[]> {
    return firstValueFrom(this.http.get<TicketMessage[]>(`/api/v1/helpdesk/tickets/${ticketId}/messages`, { params: { includeInternal: String(includeInternal) } }));
  }

  async listKbArticles(q?: string): Promise<KbArticle[]> {
    const params: Record<string, string> = {};
    if (q) params['q'] = q;
    return firstValueFrom(this.http.get<KbArticle[]>('/api/v1/kb/articles', { params }));
  }

  async createKbArticle(payload: { titleAr: string; titleEn: string; bodyAr: string; bodyEn: string; tags?: string }): Promise<KbArticle> {
    return firstValueFrom(this.http.post<KbArticle>('/api/v1/kb/articles', payload));
  }

  async publishKbArticle(id: string): Promise<KbArticle> {
    return firstValueFrom(this.http.post<KbArticle>(`/api/v1/kb/articles/${id}/publish`, {}));
  }

  async voteKbArticle(id: string, up: boolean): Promise<void> {
    await firstValueFrom(this.http.post(`/api/v1/kb/articles/${id}/vote`, { up }));
  }

  async createFromTicket(ticketId: string, titleAr: string, titleEn: string): Promise<KbArticle> {
    return firstValueFrom(this.http.post<KbArticle>(`/api/v1/kb/articles/from-ticket/${ticketId}`, { titleAr, titleEn }));
  }
}
